src/net_processing.cpp

```diff
@@ -6264,8 +6264,29 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                             }
                         }
 
-                        if (fanout || !m_txreconciliation->AddToSet(pto->GetId(), wtxid).m_succeeded) {
+                        if (fanout) {
                             vInv.push_back(inv);
+                        } else {
+                            // If the transactions fails to get into the set, we fanout
+                            auto result = m_txreconciliation->AddToSet(pto->GetId(), wtxid);
+                            if (!result.m_succeeded) {
+                                vInv.push_back(inv);
+                                // If the transaction fails because it collides with an existing one,
+                                // we also remove and fanout the conflict and all its descendants.
+                                // This is because our peer may have added the conflicting tranaction
```

Comment:
Being this part of `RelayTransaction`, which processed transactions one at a time, and given that if a collision happens the existing transaction is removed from the set, there cannot be a three-way collision. If a third transaction that happens to have the same short id as the two previous ones is added to the set later, it would be added normally since there won't be a transaction to collide with.

I could add a comment for three, but someone may then mention what happens if there are four 😅

Reply:
I was looking for something like "As a general rule, we keep the transaction that appeared first".

Currently, the `AddToSet` documentation seems outdated—it basically claims that the only way to fail is to reach a set limit.

How about the following replacement:

```
    /**
     * Step 1. Add a to-be-announced transaction to the local reconciliation set of the target peer.
     * Returns false if the set is at capacity, or if the set contains a colliding transaction.
     * Returns true if the transaction appears in the set (whether it was already there or just was added).
     */
```

Reply:
Agreed the docs for `TxReconciliationTracker::AddToSet` could be improved. I'll update it with a version of your suggestion